src/com/amazon/ion/impl/lite/IonSystemLite.java

```diff
@@ -822,4 +685,190 @@ public boolean valueIsSharedSymbolTable(IonValue value)
         return false;
     }
 
+    private class ValueLoader {
+        // This value was chosen somewhat arbitrarily; it can/should be changed if it is found to be insufficient.
+        private static final int CONTAINER_STACK_INITIAL_CAPACITY = 16;
+        private final ArrayList<IonContainerLite> containerStack;
+
+        private IonReader reader;
+
+        public ValueLoader() {
+            this.containerStack = new ArrayList<>(CONTAINER_STACK_INITIAL_CAPACITY);
+            // The reader is specified in each call to `load(IonReader)`.
+            this.reader = null;
+        }
+
+        // Does a shallow materialization of the value over which the reader is currently positioned.
+        // If the reader is positioned over a non-null container, the returned value will be an empty version
+        // of that container. Subsequent processing is required to populate it.
+        private IonValueLite shallowLoadCurrentValue() {
+            IonType ionType = reader.getType();
+            if (reader.isNullValue()) {
+                return newNull(ionType);
+            }
+
+            switch (ionType) {
+                case BOOL:
+                    return newBool(reader.booleanValue());
+                case INT:
+                    // Only construct a BigInteger if it's necessary.
+                    if (reader.getIntegerSize().equals(IntegerSize.BIG_INTEGER)) {
+                        return newInt(reader.bigIntegerValue());
+                    }
+                    return newInt(reader.longValue());
```

Comment:
This could end up being a pretty nice efficiency improvement over the previous implementation...